autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+import logging
+import time
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import numpy as np
+
+import torch
+from torch.autograd import Variable
+from torch.utils.tensorboard.writer import SummaryWriter
+
+from autoPyTorch.pipeline.components.training.base_training import autoPyTorchTrainingComponent
+
+
+class BudgetTracker(object):
+    def __init__(self,
+                 budget_type: str,
+                 max_value: int,
+                 ):
+        """
+        An object for tracking when to stop the network training.
+        It handles epoch based criteria as well as training based criteria.
+
+
+        # Maybe it is better to use pynisher? Remember the whole everything gets
+        copied when forking memory, so it makes sense to ask... but also,
+        how to we regulate memory from GPU and CPU?
+        """
+        self.start_time = time.time()
+        self.budget_type = budget_type
+        self.max_value = max_value
+
+    def is_max_epoch_reached(self, epoch: int) -> bool:
+        if self.budget_type == 'epochs' and epoch > self.max_value:
+            return True
+        return False
+
+    def is_max_time_reached(self) -> bool:
+        elapsed_time = time.time() - self.start_time
+        if self.budget_type == 'runtime' and elapsed_time > self.max_value:
+            return True
+        return False
+
+
+class RunSummary(object):
```

Comment:
Maybe add a description that this is done epoch wise

Reply:
I have added more documentation to this class